File: Java/src/gov/nasa/larcfm/Util/Debug.java
/*
 * Copyright (c) 2013-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.Util;

public class Debug {
	private static boolean userDebugFlag = false;

	public static void setUserDebugFlag(boolean b) {
		userDebugFlag = b;
	}

	/**
	 * Print out a debugging message msg, with each line prepended with tag
	 * 
	 * @param tag the tag to indicate the location of this debug message.
	 * @param msg the debug message
	 */
	public static void pln(String tag, String msg) {
		if (userDebugFlag) {
			String[] lines = msg.split("\\n");
			for (int i = 0; i < lines.length; i++) {
				System.out.println("<!! "+tag+" !!> "+lines[i]);
			}
		}
	}

	/**
	 * Print out a debugging message.  Be careful with this method, it may be hard to 
	 * find these messages in the future, use pln(tag,msg) instead.
	 * 
	 * @param msg the debug message
	 */
	public static void pln(String msg) {
		pln("empty",msg);
	}

	public static void halt(String str) {
		System.out.println(str);
		if (Constants.allowDebugHalt) {
			Thread.dumpStack();
			System.exit(1);
		}
	}

	public static void halt() {
		halt("");	
	}
	

}

File: C++/src/ACCoRD/Debug.cpp
/*
 * Copyright (c) 2011-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "Debug.h"
#include "string_util.h"
#include <stdlib.h>
#include <string>
#include <iostream>

#ifdef __GNUC__
#include <execinfo.h>
#endif


namespace larcfm {
  
  using std::string;
  using std::endl;
  using std::cout;

  bool Debug::userDebugFlag = false;

  void Debug::setUserDebugFlag(bool b) {
	  userDebugFlag = b;
  }

  void Debug::pln(const string& tag, const string& msg) {
	  if (userDebugFlag) {
			std::vector<string> lines = split(msg,"\n");
			for (unsigned int i = 0; i < lines.size(); i++) {
				cout << "<!! " << tag << " !!> " << lines[i] << endl;
			}
			// no msg string
			if (lines.size() == 0) {
				cout << "<!! " << tag << " !!> " << endl;
			}
	  }
  }
  
  void Debug::halt() {
#ifdef __GNUC__
	  cout << "Debug::halt() called.  Compiling with the DEBUG=-g option _may_ make outputs more readable" << endl;
	  int sz = 10;
	  size_t size;
	  void *array[sz];
	  size = backtrace(array, sz);
	  char **strings = backtrace_symbols(array,size);
	  for (int i = 0; i < sz; i++) {
		  cout << strings[i] << endl;
	  }
#else
	  cout << "Debug::halt() called.  Backtrace is a glibc (gcc) function" << endl;
#endif
	  exit(1);
 }



//  void Debug::pln(const string& msg)
//  {
//#ifdef __GLIBC__
//      void *array[4];
//      size_t size;
//      char **strings;
//      size_t i;
//
//      size = backtrace (array, 4);
//      strings = backtrace_symbols (array, size);
//
//      if (size > 2) {
//    	  Debug::pln(strings[2],msg);
//      } else {
//    	  Debug::pln("",msg);
//      }
//      free (strings);
//#else
//        Debug::pln("",msg);
//#endif
//  }

}
